kotlin/workflow-runtime/src/main/java/com/squareup/workflow/diagnostic/WorkflowDiagnosticListener.kt

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright 2019 Square Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.squareup.workflow.diagnostic
+
+import com.squareup.workflow.VeryExperimentalWorkflow
+import com.squareup.workflow.WorkflowAction
+import kotlinx.coroutines.CoroutineScope
+
+/**
+ * Receives diagnostic events from the workflow runtime when significant things happen, e.g. the
+ * runtime starts or stops, workflows and workers are started or stopped, and information about
+ * render passes and workflow actions.
+ *
+ * To register a listener with the runtime, set
+ * [WorkflowSession.diagnosticListener][com.squareup.workflow.WorkflowSession.diagnosticListener]
+ * in the `beforeStart` function passed to
+ * [launchWorkflowIn][com.squareup.workflow.launchWorkflowIn].
+ *
+ * Multiple listeners can be composed using the [andThen] function.
+ *
+ * No guarantees are made about what threads methods are invoked on. Implementations should take
+ * care to be thread-safe.
+ *
+ * ## Detekt Configuration
```

Comment:
Seems like an odd thing to document.

Reply:
Agreed, but not sure what a better alternative would be.